src/qcrypto/simbasics.py

```diff
@@ -1,132 +1,424 @@
-        # Recipient obtains the qubits from self and measures using given basis
-        for qubit, base in zip(self.qubits, recipient_bases):
-            recipient.measure(qubit.copy(), base)
+        self._state = np.array(new_state)
+        self._normalize_state()
 
-        # Recipient construct qubits based on these measurements and on their own basis selection
-        recipient.qubits = []
-        for base, measurement in zip(recipient_bases, recipient.measurements):
-            received_qubit = Qubit(measurement * np.pi, base)  # measurement = 0 or 1
-            recipient.qubits.append(received_qubit)
-        recipient.qubits = np.array(recipient.qubits)
+    def _normalize_state(self):
+        """
+        Updates the state of the system by normalizing its states.
 
-    def get_key(self, bases):
-        for qubit, base in zip(self.qubits, bases):
-            measurement = self.measure(qubit, base)
-            self.key.append(measurement)
-        return self.key
+        Args:
+            n/a
 
-    def send_classic(self, receiverAgent, bits):
-        pass
+        Retuns:
+            n/a
+        """
+        self._state /= np.linalg.norm(self._state)
+
+    def apply_gate(self, gate: npt.NDArray[np.complex_]):
+        """
+        Applies quantum gate to the system of qubits.
+
+        Args:
+            gate (np.NDarray): Gate to be applied to the to the system
+
+        Returns:
+            n/a
 
-    def genkey(self, numqubits, numcheckbits):
         """
-        Generates an array of qubits
+        N = int(np.log2(len(self._state)))
+        gate = tensor_power(gate, N)
+        self._state = np.dot(gate, self._state)
+        self._normalize_state()
+
+    def __str__(self):
+        return str(self._state)
+
+    def __repr__(self):
+        return str(self._state)
+
+
+@dataclasses.dataclass
+class Agent:
+    num_priv_qubits: int = None
+    qstates: Dict[str, npt.NDArray[np.complex_]] = dataclasses.field(
+        default_factory=lambda: {"private": None, "public": None}
+    )
+    keys: Dict[str, npt.NDArray[np.int_]] = dataclasses.field(
+        default_factory=lambda: {"private": None, "public": None}
+    )
+    priv_qstates: Union[QstateEnt, QstateUnEnt] = None
+    init_method: str = "random"
+    priv_qbittype: str = None
+
+    def __post_init__(self):
+        self.qstates = {"private": None, "public": None}
+
+        if self.priv_qstates is None and self.priv_qbittype == "entangled":
+            self.set_qstate(
+                QstateEnt(
+                    num_qubits=self.num_priv_qubits, init_method=self.init_method
+                ),
+                "private",
+            )
+        elif self.priv_qstates is None and self.priv_qbittype == "unentangled":
+            self.set_qstate(
+                QstateUnEnt(
+                    num_qubits=self.num_priv_qubits, init_method=self.init_method
+                ),
+                "private",
+            )
+
```

Comment:
I'm assuming Python does the correct thing here resolving the order of operations as desired on lines 362 and 369 but probably should add `()` around those `<condition> and <condition> `

Reply:
It does the correct thing, but you are right: for clarity, we should add some parentheses.